dashboard.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+import collections
+import glob
+import importlib
+import inspect
+import os
+import pathlib
+import sys
+
+import altair
+import pandas
+import panel
+import sqlalchemy
+from bokeh.resources import INLINE
+
+panel.extension("vega")
+
+
+def get_test_source():
+    """
+    Crawl the tests directory and try to grab code for each test on a best-effort
+    basis. This relies on the tests being importable from this script, so the
+    environment should be similar enough that that is possible.
+    """
+    source: dict[str, str] = {}
+    files = glob.glob("tests/**/test_*.py", recursive=True)
+    for f in files:
+        try:
+            # Fragile!
```

Comment:
Does this mean that it sometimes works and some doesn't?

Reply:
It works today, but it might fail if our test suite includes other dependencies, has weird side effects, or otherwise makes it difficult to import. A nicer approach might be to parse it with `ast` in some way, but I didn't want to go too far down a rabbit hole, and `inspect` is fairly easy to use.

I'm trying to be fairly forgiving here (with a wide try/except), so I think the worst case scenario is that the code is missing for some (or all) functions.